cmd/util/ledger/migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,261 @@
+		// dont fail fast... try to collect errors so multiple errors can be addressed at once
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		migratedPayloads = append(migratedPayloads, results...)
+	}
+	if len(errors) != 0 {
+		return nil, &MultipleContractMigrationError{
+			Errors: errors,
+		}
+	}
+	return migratedPayloads, nil
+}
+
+func keyToRegisterId(key ledger.Key) (flow.RegisterID, error) {
+	if len(key.KeyParts) != 3 ||
+		key.KeyParts[0].Type != state.KeyPartOwner ||
+		key.KeyParts[1].Type != state.KeyPartController ||
+		key.KeyParts[2].Type != state.KeyPartKey {
+		return flow.RegisterID{}, fmt.Errorf("key not in expected format %s", key.String())
+	}
+
+	return flow.NewRegisterID(string(key.KeyParts[0].Value), string(key.KeyParts[1].Value), string(key.KeyParts[2].Value)), nil
+}
+
+func migrateRegister(p ledger.Payload, cache map[string]struct{}, logger zerolog.Logger) ([]ledger.Payload, error) {
+	registerId, err := keyToRegisterId(p.Key)
+	if err != nil {
+		return nil, err
+	}
+	// have we migrated this?
+	if _, alreadyMigrated := cache[registerId.Owner]; alreadyMigrated {
+		// yes, return it as is
+		return []ledger.Payload{p}, nil
+	}
+	if !needsMigration(registerId) {
+		return []ledger.Payload{p}, nil
+	}
+
+	// we are going to migrate this
+	cache[registerId.Owner] = struct{}{}
+
+	if em, hasEM := MultipleContractsSpecialMigrations[registerId.Owner]; hasEM {
+		logger.Info().
+			Err(err).
+			Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+			Msg("Using exceptional migration for address")
+		return em(p, logger)
+	}
+
+	return migrateValue(p, logger)
+}
+
+func migrateValue(p ledger.Payload, l zerolog.Logger) ([]ledger.Payload, error) {
+
+	// we don't need the the empty code register
+	value := p.Value
+	address := flow.BytesToAddress(p.Key.KeyParts[0].Value)
+	if len(value) == 0 {
+		return []ledger.Payload{}, nil
+	}
+
+	code := string(value)
+	program, err := parser2.ParseProgram(code)
+	if err != nil {
+		l.Error().
+			Err(err).
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Cannot parse program at address")
+		return nil, err
+	}
+	declarations := program.Declarations
+
+	// find import declarations
+	importDeclarations := make([]ast.Declaration, 0)
+	for _, d := range declarations {
+		if _, isImport := d.(*ast.ImportDeclaration); isImport {
+			importDeclarations = append(importDeclarations, d)
+		}
+	}
+
+	// sort them backwards because we will be slicing the code string
+	sort.SliceStable(importDeclarations, func(i, j int) bool {
+		return importDeclarations[i].StartPosition().Offset > importDeclarations[j].StartPosition().Offset
+	})
+
+	imports := ""
+	for _, d := range importDeclarations {
+		// aggregate imports for later use
+		imports = imports + code[d.StartPosition().Offset:d.EndPosition().Offset+1] + "\n"
+		// remove imports from the code
+		code = code[:d.StartPosition().Offset] + code[d.EndPosition().Offset+1:]
```

Comment:
Maybe refactor `d.StartPosition().Offset` and `d.EndPosition().Offset` out and declare variables for them, so the slicing expressions become easier to read